src/EventStore.Core/Messaging/RequestResponseAltDispatcher.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using EventStore.Core.Bus;
+using EventStore.Core.Helpers;
+using EventStore.Core.Messages;
+
+namespace EventStore.Core.Messaging {
+	// Just a typedef
+	public class ReadDispatcher :
+		RequestResponseAltDispatcher<
+			ClientMessage.ReadStreamEventsBackward,
+			ClientMessage.ReadStreamEventsBackwardCompleted,
+			ClientMessage.NotHandled,
+			IReadStreamEventsBackwardHandler> {
+		public ReadDispatcher(
+			IPublisher publisher,
+			Func<ClientMessage.ReadStreamEventsBackward, Guid> getRequestCorrelationId,
+			Func<ClientMessage.ReadStreamEventsBackwardCompleted, Guid> getResponseCorrelationId,
+			Func<ClientMessage.NotHandled, Guid> getResponseAltCorrelationId,
+			IEnvelope defaultReplyEnvelope,
+			Func<Guid, Message> cancelMessageFactory = null) :
+				base(publisher, getRequestCorrelationId, getResponseCorrelationId,
+					getResponseAltCorrelationId, defaultReplyEnvelope, cancelMessageFactory) {
+		}
+	}
+
+	// This derivation of RequestResponseDispatcher handles two different response messages
+	// Useful if the system might respond to the request with a message like NotHandled and we want to know about it.
+	public class RequestResponseAltDispatcher<TRequest, TResponse, TResponseAlt, THandler> :
+		RequestResponseDispatcher<TRequest, TResponse, THandler>,
+		IHandle<TResponseAlt>
+		where TRequest : Message
+		where TResponse : Message
+		where TResponseAlt : Message
+		where THandler : IHandle<TResponse>, IHandleAlt<TResponseAlt>, IHandleTimeout {
+
+		private readonly Func<TResponseAlt, Guid> _getResponseAltCorrelationId;
+
+		public RequestResponseAltDispatcher(
+			IPublisher publisher,
+			Func<TRequest, Guid> getRequestCorrelationId,
+			Func<TResponse, Guid> getResponseCorrelationId,
+			Func<TResponseAlt, Guid> getResponseAltCorrelationId,
```

Comment:
i'm not sure i understand why there's a need for the response correlation IDs, don't they always have the same value as the request correlation id?

Reply:
ahh ok got it, i thought we knew the request correlation id upfront, but it's generated only after the read request is sent (otherwise we could just pass in a single id here instead of 3 functions)